### PR TITLE
SO-2902: Add multi-language RF2 import support.

### PR DESCRIPTION
**Pull request by zoliMolnar**
This pull request enables multi-language RF2 importing.

### CODE REVIEW

File: snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java [hunk 1]
```diff
@@ -203,5 +207,52 @@ public void import11ExtensionConceptWithVersion() {
 		getComponent(branchPath, SnomedComponentType.CONCEPT, "555231000005107").statusCode(200);
 		getVersion("SNOMEDCT-NE", "2015-02-05").statusCode(200);
 	}
-
+	
+	@Test
+	public void importMoreThanOneLanguageCodeDescription() {
+		final String enDescriptionId = "41320138114";
+		final String svDescriptionId = "24688171113";
+		final String conceptIdODescription = "301795004";
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java]: Typo

File: snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java [hunk 2]
```diff
@@ -203,5 +207,52 @@ public void import11ExtensionConceptWithVersion() {
 		getComponent(branchPath, SnomedComponentType.CONCEPT, "555231000005107").statusCode(200);
 		getVersion("SNOMEDCT-NE", "2015-02-05").statusCode(200);
 	}
-
+	
+	@Test
+	public void importMoreThanOneLanguageCodeDescription() {
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java]: Please add a number to the test method's name like the others above. As you can see there is an annotation at the top: @FixMethodOrder(MethodSorters.NAME_ASCENDING) which enforces method execution order.

File: snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java [hunk 3]
```diff
@@ -203,5 +207,52 @@ public void import11ExtensionConceptWithVersion() {
...
+		getComponent(branchPath, SnomedComponentType.CONCEPT, conceptIdODescription).statusCode(200);
+		
+		importArchive("SnomedCT_Release_INT_20150201_new_description_with2_language_code.zip", importConfiguration);
+		getComponent(branchPath, SnomedComponentType.DESCRIPTION, enDescriptionId).statusCode(200);
+		getComponent(branchPath, SnomedComponentType.DESCRIPTION, svDescriptionId).statusCode(200);
+	}
+	
+	@Test
+	public void importFromMoreThanOneLanguageRefset() {
+		final String enLanguageRefsetConceptId = "34d07985-48a0-41e7-b6ec-b28e6b00adfc";
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java]: You should decide to either merge the two test cases together (since the RF2 files are the same and you always import both descriptions and language refset members as well) or split the RF2 files and test language refset and description import separately.

File: snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java (hunk 2 above)
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java]: importDescriptionsWithMultipleLanguageCodes or importDescriptionsWithDifferentLanguageCodes

File: snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java [hunk 4]
```diff
@@ -203,5 +207,52 @@ public void import11ExtensionConceptWithVersion() {
...
+		importArchive("SnomedCT_Release_INT_20150131_new_concept_for_languageCode.zip");
+		getComponent(branchPath, SnomedComponentType.CONCEPT, conceptIdODescription).statusCode(200);
+		
+		importArchive("SnomedCT_Release_INT_20150201_new_description_with2_language_code.zip", importConfiguration);
+		getComponent(branchPath, SnomedComponentType.DESCRIPTION, enDescriptionId).statusCode(200);
+		getComponent(branchPath, SnomedComponentType.DESCRIPTION, svDescriptionId).statusCode(200);
+	}
+	
+	@Test
+	public void importFromMoreThanOneLanguageRefset() {
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java]: You are not importing more than one language refset! It should be rather something like:
importLanguageRefsetMembersWithMultipleLanguageCodes or importLanguageRefsetMembersWithDifferentLanguageCodes

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java [hunk 5]
```diff
@@ -107,13 +107,20 @@ public void execute(final CommandInterpreter interpreter) {
 			// Setting up configuration only with the required fields
 			config.setSourceKind(ImportSourceKind.ARCHIVE);
 			config.setArchiveFile(archiveFile);
-			config.setDescriptionsFile(new File(archiveFileSet.getFileName(zipFiles, ReleaseComponentType.DESCRIPTION, contentSubType)));
-			config.setLanguageRefSetFile(languageRefSetFile);
+			final Collection<String> allFileName = archiveFileSet.getAllFileName(zipFiles, ReleaseComponentType.DESCRIPTION, contentSubType);
+			final Collection<File> descriptionFileNames = Collections.emptySet();
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java]: Please fix the initialization of this set. You create an immutable set and then try to add elements to it??
Here is the javadoc of Collections.emptySet(): "Returns an empty set (immutable)." Javadoc is your friend! Nevertheless this could be solved with Java 8 streams in one line.

File: snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java [hunk 6]
```diff
@@ -131,14 +134,18 @@ public void setRelationshipsFile(final File relationshipsFile) {
...
 
-	public void setLanguageRefSetFile(final File languageRefSetFile) {
-		this.languageRefSetFile = languageRefSetFile;
+	public void setLanguageRefSetFiles(final Collection<File> languageRefSetFiles) {
+		this.languageRefSetFiles = Collections3.toImmutableSet(languageRefSetFiles);
 	}
-
+	
+	public void addLanguageRefSetFiles(final File languageRefSetFile) {
+		this.languageRefSetFiles.add(languageRefSetFile);
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java]: What happens if I call this add method right after I called setLanguageRefSetFiles(Collection<File> files) ????

File: snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java [hunk 7]
```diff
@@ -131,14 +134,18 @@ public void setRelationshipsFile(final File relationshipsFile) {
...
 	}
 
-	public void setLanguageRefSetFile(final File languageRefSetFile) {
-		this.languageRefSetFile = languageRefSetFile;
+	public void setLanguageRefSetFiles(final Collection<File> languageRefSetFiles) {
+		this.languageRefSetFiles = Collections3.toImmutableSet(languageRefSetFiles);
 	}
-
+	
+	public void addLanguageRefSetFiles(final File languageRefSetFile) {
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java]: If I would like to add **one** language refset file to a collection why the method name addLanguageRefSetFile_**s**_?

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java [hunk 8]
```diff
@@ -107,13 +107,20 @@ public void execute(final CommandInterpreter interpreter) {
...
+			for (String fileName : allFileName) {
+				descriptionFileNames.add(new File(fileName));
+			}
+			config.setDescriptionsFiles(descriptionFileNames);
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java]: We are in a for loop here, why do you add all the description files to the config in each iteration?

File: snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java [hunk 9]
```diff
@@ -58,10 +61,10 @@
 	private File archiveFile;
 	private File rootFile;
 	private File conceptsFile;
-	private File descriptionsFile;
+	private Collection<File> descriptionsFiles = Collections.emptySet();
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java]: Please fix the initialization of this collection

File: snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java [hunk 10]
```diff
@@ -58,10 +61,10 @@
...
 	private File relationshipsFile;
 	private File statedRelationshipsFile;
-	private File languageRefSetFile;
+	private Collection<File> languageRefSetFiles = Collections.emptySet();
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java]: Please fix the initialization of this collection

File: snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java [hunk 11]
```diff
@@ -18,13 +18,16 @@
 import java.io.File;
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java:20]: I get the following exception when I open the import wizard in the client (server-client mode): 
`java.lang.IllegalArgumentException: Could not find property with name descriptionsFile in class class com.b2international.snowowl.snomed.importer.net4j.ImportConfiguration
	at org.eclipse.core.internal.databinding.beans.BeanPropertyHelper.getPropertyDescriptor(BeanPropertyHelper.java:168)
	at org.eclipse.core.databinding.beans.PojoProperties.value(PojoProperties.java:119)
	at org.eclipse.core.databinding.beans.PojoProperties.value(PojoProperties.java:89)
	at org.eclipse.core.databinding.beans.PojoObservables.observeValue(PojoObservables.java:76)
	at org.eclipse.core.databinding.beans.PojoObservables.observeValue(PojoObservables.java:56)
	at com.b2international.snowowl.snomed.importer.ui.wizard.page.ImportDetailsWizardPage.bindTextField(ImportDetailsWizardPage.java:319)
	at com.b2international.snowowl.snomed.importer.ui.wizard.page.ImportDetailsWizardPage.initDataBinding(ImportDetailsWizardPage.java:236)
	at com.b2international.snowowl.snomed.importer.ui.wizard.page.AbstractImportWizardPage.createControl(AbstractImportWizardPage.java:67)`

File: snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/SnomedImportRequest.java [hunk 12]
```diff
@@ -69,16 +70,23 @@ protected void requesting(final ExtendedDataOutputStream out, final OMMonitor mo
 			}
 			
 			out.writeUTF(importConfiguration.getCodeSystemShortName());
-			
+			final Collection<File> descriptionsFiles = importConfiguration.getDescriptionsFiles();
+			final Collection<File> languageRefSetFiles = importConfiguration.getLanguageRefSetFiles();
 			monitor.worked(); // 1
 			
+			for (File descfile : descriptionsFiles) {
+				writeComponent(out, descfile, monitor);
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/SnomedImportRequest.java]: This is not going to work like this, see how refset URLs or excluded refset ids are sent over the wire.

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/net4j/SnomedImportIndication.java [hunk 13]
```diff
@@ -92,12 +93,12 @@ protected void indicating(final ExtendedDataInputStream in, final OMMonitor moni
 			receivedFilesDirectory.deleteOnExit();
 			
 			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setConceptsFile(f); }}, monitor.fork());
-			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setDescriptionsFile(f); }}, monitor.fork());
+			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setDescriptionsFiles(ImmutableList.of(f)); }}, monitor.fork());
 			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setTextDefinitionFile(f); }}, monitor.fork());
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/net4j/SnomedImportIndication.java]: This is not going to work like this, see how refset URLs or excluded refset ids are sent over the wire.

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java [hunk 14]
```diff
@@ -173,11 +173,15 @@ public void handleRecord(int recordCount, List<String> record) {
 					subMonitor.setWorkRemaining(LARGE_WORK_REMAINING);
 				}
 			};
-
-			URL url = configuration.toURL(configuration.getDescriptionsFile());
-			descriptionReader = new InputStreamReader(url.openStream());
-			CsvParser parser = new CsvParser(descriptionReader, getFileName(url), CSV_SETTINGS, descriptionParserCallback, DESCRIPTION_FIELD_COUNT);
-			parser.parse();
+			if (!configuration.getDescriptionsFiles().isEmpty()) {
+				for(File descFile : configuration.getDescriptionsFiles()) {
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java]: missing space

File: snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java [hunk 15]
```diff
@@ -203,5 +207,52 @@ public void import11ExtensionConceptWithVersion() {
...
+		final Map<?, ?> importConfiguration = ImmutableMap.builder()
+				.put("type", Rf2ReleaseType.DELTA.name())
+				.put("branchPath",  branchPath.getPath())
+				.put("createVersions", false)
+				.build();
+		
+		importArchive("SnomedCT_Release_INT_20150131_new_concept_for_languageCode.zip");
+		getComponent(branchPath, SnomedComponentType.CONCEPT, conceptIdODescription).statusCode(200);
+		
+		importArchive("SnomedCT_Release_INT_20150201_new_description_with2_language_code.zip", importConfiguration);
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java]: The description in sct2_Description_Delta-sv_INT_20150201 has a language code 'en', please fix that!

**Review by nagyo — CHANGES_REQUESTED** [review]

File: snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java [hunk 16]
```diff
@@ -203,5 +207,36 @@ public void import11ExtensionConceptWithVersion() {
...
+	
+	@Test
+	public void import12WithMultipleLanguageCodes() {
+		final String enDescriptionId = "41320138114";
+		final String svDescriptionId = "24688171113";
+		final String enLanguageRefsetMemberId = "34d07985-48a0-41e7-b6ec-b28e6b00adfc";
+		final String svLanguageRefsetMemberId = "34d07985-48a0-41e7-b6ec-b28e6b00adfb";
+		final String conceptIdOfDescription = "301795004";
+		
+
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/io/SnomedImportApiTest.java]: Unnecessary extra line

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java [hunk 17]
```diff
@@ -93,27 +93,40 @@ public void execute(final CommandInterpreter interpreter) {
...
 			config.setSourceKind(ImportSourceKind.ARCHIVE);
 			config.setArchiveFile(archiveFile);
-			config.setDescriptionsFile(new File(archiveFileSet.getFileName(zipFiles, ReleaseComponentType.DESCRIPTION, contentSubType)));
-			config.setLanguageRefSetFile(languageRefSetFile);
 	
 			final SnomedRefSetNameCollector provider = new SnomedRefSetNameCollector(config, new NullProgressMonitor(), "");
 	
 			try {
-				provider.parse(config.toURL(config.getLanguageRefSetFile()));
+				for (File langFile : config.getLanguageRefSetFiles()) {
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java]: You are already iterating through the available language refset files in the outer for loop, I don't think this is necessary here.

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java [hunk 18]
```diff
@@ -173,11 +173,15 @@ public void handleRecord(int recordCount, List<String> record) {
...
-			descriptionReader = new InputStreamReader(url.openStream());
-			CsvParser parser = new CsvParser(descriptionReader, getFileName(url), CSV_SETTINGS, descriptionParserCallback, DESCRIPTION_FIELD_COUNT);
-			parser.parse();
+			if (!configuration.getDescriptionsFiles().isEmpty()) {
+				for (File descFile : configuration.getDescriptionsFiles()) {
+					final URL url = configuration.toURL(descFile);
+					descriptionReader = new InputStreamReader(url.openStream());
+					final CsvParser parser = new CsvParser(descriptionReader, getFileName(url), CSV_SETTINGS, descriptionParserCallback, DESCRIPTION_FIELD_COUNT);
+					parser.parse();
+
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java]: Extra line

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/RF2ReleaseRefSetFileCollector.java [hunk 19]
```diff
@@ -67,9 +67,10 @@
 		// Add reference set URLs from the first wizard page in case they're not present
 		try {
 			
-			if (null != configuration.getLanguageRefSetFile()) {
-				collectedUrlSet.add(configuration.toURL(configuration.getLanguageRefSetFile()));
+			for (File langFiles : configuration.getLanguageRefSetFiles()) {
+				collectedUrlSet.add(configuration.toURL(langFiles));
 			}
+
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/RF2ReleaseRefSetFileCollector.java]: Extra line

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/validation/SnomedValidationContext.java [hunk 20]
```diff
@@ -162,9 +164,12 @@ private void addRefSetFilesForValidating() throws IOException {
...
 		if (!Iterables.any(releaseFileValidators, Predicates.instanceOf(SnomedLanguageRefSetValidator.class))) {
-			
-			if (isValidReleaseFile(configuration.getLanguageRefSetFile())) {
-				releaseFileValidators.add(new SnomedLanguageRefSetValidator(configuration, configuration.toURL(configuration.getLanguageRefSetFile()), this));
+			Collection<File> languageRefSetFiles = configuration.getLanguageRefSetFiles();
+			for (File langFile : languageRefSetFiles) {
+				if (isValidReleaseFile(langFile)) {
+					releaseFileValidators.add(new SnomedLanguageRefSetValidator(configuration, configuration.toURL(langFile), this));
+				}
+
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/validation/SnomedValidationContext.java]: Extra line

File: snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java (hunk 9 above)
Comment (nagyo) [reply snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/ImportConfiguration.java]: Thanks to this the collection will be initialized still as immutable.

File: snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/SnomedImportRequest.java [hunk 21]
```diff
@@ -70,15 +71,27 @@ protected void requesting(final ExtendedDataOutputStream out, final OMMonitor mo
 			
 			out.writeUTF(importConfiguration.getCodeSystemShortName());
 			
+			final Collection<File> descriptionsFiles = importConfiguration.getDescriptionsFiles();
+			final Collection<File> languageRefSetFiles = importConfiguration.getLanguageRefSetFiles();
 			monitor.worked(); // 1
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/SnomedImportRequest.java]: Why is this here? Have you checked how the monitor's work is calculated here?

File: snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/SnomedImportRequest.java [hunk 22]
```diff
@@ -70,15 +71,27 @@ protected void requesting(final ExtendedDataOutputStream out, final OMMonitor mo
...
+				writeComponent(out, langFile, monitor.fork());
+			}
+			
 			writeComponent(out, importConfiguration.getConceptsFile(), monitor.fork());
-			writeComponent(out, importConfiguration.getDescriptionsFile(), monitor.fork());
 			writeComponent(out, importConfiguration.getTextDefinitionFile(), monitor.fork());
 			writeComponent(out, importConfiguration.getRelationshipsFile(), monitor.fork());
 			writeComponent(out, importConfiguration.getStatedRelationshipsFile(), monitor.fork());
 			writeComponent(out, importConfiguration.getDescriptionType(), monitor.fork());
-			writeComponent(out, importConfiguration.getLanguageRefSetFile(), monitor.fork()); // + 7
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer/src/com/b2international/snowowl/snomed/importer/net4j/SnomedImportRequest.java]: That comment is a clue!

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/net4j/SnomedImportIndication.java [hunk 23]
```diff
@@ -87,17 +89,44 @@ protected void indicating(final ExtendedDataInputStream in, final OMMonitor moni
...
 			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setRelationshipsFile(f); }}, monitor.fork());
 			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setStatedRelationshipsFile(f); }}, monitor.fork());
 			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setDescriptionType(f); }}, monitor.fork());
-			readComponent(in, importConfiguration, receivedFilesDirectory, new FileCallback() { @Override public void setFile(final File f) { importConfiguration.setLanguageRefSetFile(f); }}, monitor.fork());
+
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/net4j/SnomedImportIndication.java]: Extra line

**Review by nagyo — CHANGES_REQUESTED** [review]
I just realized that multiple Text Definition files could also be present in RF2 archives based on the language code... Please add that feature as well. They work just as the same as description files do.

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java [hunk 24]
```diff
@@ -93,52 +94,56 @@ public void execute(final CommandInterpreter interpreter) {
...
+		}).collect(Collectors.toSet());
+		
+		final SnomedRefSetNameCollector provider = new SnomedRefSetNameCollector(refsetUrls, config, new NullProgressMonitor());
+		provider.parse();
+		
+		// Setting up configuration only with the required fields
+		config.setSourceKind(ImportSourceKind.ARCHIVE);
+		config.setArchiveFile(archiveFile);
+		
+		for (final Entry<String, String> label : provider.getRefsetIdToLabelMap().entrySet()) {
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java:131]: This label map could also be extended with information from the store (from MAIN, most likely), similar to how it's done in the import wizard.

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java [hunk 25]
```diff
@@ -67,210 +73,219 @@
...
-			put(ErroneousAustralianReleaseFileNames.ERRONEOUS_AU_20120229_SUBPACK_QUANTITY_FULL_REFSET_NAME, Concepts.REFSET_CONCRETE_DOMAIN_TYPE_AU).
-			put(ErroneousAustralianReleaseFileNames.ERRONEOUS_AU_20120229_UNIT_OF_USE_QUANTITY_REFSET_NAME, Concepts.REFSET_CONCRETE_DOMAIN_TYPE_AU).
-			put(ErroneousAustralianReleaseFileNames.ERRONEOUS_AU_20120229_UNIT_OF_USE_SIZE_REFSET_NAME, Concepts.REFSET_CONCRETE_DOMAIN_TYPE_AU).
-			put(ErroneousAustralianReleaseFileNames.ERRONEOUS_AU_20110531_CTV3_ID_REFSET_NAME, Concepts.REFSET_SIMPLE_MAP_TYPE).
-			put(ErroneousAustralianReleaseFileNames.ERRONEOUS_AU_20110531_SNOMED_RT_REFSET_NAME, Concepts.REFSET_SIMPLE_MAP_TYPE).
-			put(ErroneousAustralianReleaseFileNames.ERRONEOUS_AU_20110531_SNOMED_LANGUAGE_REFSET_NAME, Concepts.REFSET_LANGUAGE_TYPE).
-			put(ErroneousAustralianReleaseFileNames.ERRONEOUS_AU_20120229_SNOMED_LANGUAGE_REFSET_NAME, Concepts.REFSET_LANGUAGE_TYPE).build();
+	private static final String UNLABELED_REFSET_SUFFIX = " (unresolved)";
+	private static final String ESCAPED_UNLABELED_REFSET_SUFFIX = CharMatcher.is(')')
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java]: `Pattern#quote(String)` will return a String that is treated as a "literal pattern" in regular expressions; suggesting to use that instead of targeted replacements.

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java [hunk 26]
```diff
@@ -67,210 +73,219 @@
...
 				
-				@Override
-				public void handleRecord(int recordCount, List<String> record) {
+				if (!unlabeledRefSetIds.isEmpty()) {
 					
-					if (recordCount == 1) {
-						refSetTypeIconId = getRefSetTypeIconIdFromHeader(url, record);
-					} else {					
-						unlabeledRefSetIds.add(record.get(CONCEPT_ID_COLUMN));
+					InputStreamReader descriptionReader = null;
```
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/util/SnomedRefSetNameCollector.java]: `descriptionReader` should get the try-with-resources treatment:

```java

final URL url = configuration.toURL(descFile);
try (final InputStreamReader descriptionReader = new InputStreamReader(url.openStream())) {
    ...
}
```

**Review by apeteri — CHANGES_REQUESTED** [review]
More copyright dates to move forward:

* `SnomedImportApiTest`
* `ImportRefSetCommand`
* `ListLanguageRefSetsCommand`
* `ImportUtil`
* `RF2ReleaseRefSetFileCollector`
* `SnomedRefSetNameCollector`
* `AbstractSnomedValidator`
* `SnomedConceptValidator`
* `SnomedTaxonomyValidator`
* `SnomedValidationContext`
* `ImportConfiguration`
* `ReleaseFileSet`
* `ReleaseFileSetSelectors`

File: snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java (hunk 24 above)
Comment (apeteri) [reply snomed/com.b2international.snowowl.snomed.importer.rf2/src/com/b2international/snowowl/snomed/importer/rf2/command/ListLanguageRefSetsCommand.java:131]: Sorry, I see that the name collector does this already. ✔️

**Review by apeteri — APPROVED** [review]

**Review by nagyo — APPROVED** [review]

**Review by cmark — APPROVED** [review]
[LGTM](https://www.lgtm.in/i/YO8oFNJLs)
